### PR TITLE
[inductor] Vendor and adapt QuACK symmetric GEMM

### PR DESCRIPTION
**Pull request by mlazos**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #194181
* #194180
* #194179
* #194178
* __->__ #194177

Human commentary from the author: "and add architecture descriptions and examples to each PR to make it clear"

> AI-assisted change description:
>
> ## Summary
>
> A product such as `X @ X.T` is symmetric, so a conventional GEMM computes
> two equal output triangles. Vendor the existing QuACK symmetric GEMM from the
> pinned upstream revision and adapt it to PyTorch runtime interfaces. The kernel
> schedules one triangle and writes each result to both symmetric locations. It
> supports contiguous FP16 and BF16 matrices, homogeneous batches, and an optional
> symmetric `C` epilogue.
>
> ## Architecture
>
> The vendoring script copies `quack/gemm_symmetric.py` from QuACK commit
> `99bd7973bf3dc6db40961e413d4bdfea6c6fee3e` and applies the PyTorch adaptation
> in `0015-inductor-symmetric-gemm.patch`. The resulting kernel composes existing
> QuACK building blocks rather than defining another GEMM mainloop:
>
> ```text
> GemmSm100 tensor-core mainloop
>          +
> GemmSymmetricMixin
>   - selects TriangularTileScheduler
>   - exposes the transposed output view to the epilogue
>          =
> GemmSymmetricSm100
> ```
>
> `TriangularTileScheduler` assigns only tiles on one side of the diagonal.
> The normal SM100 TMA, MMA, pipeline, and cluster implementation remains
> unchanged. The epilogue stores the computed tile through both the ordinary
> and transposed output views, producing a dense symmetric result without a
> second GEMM.
>
> The compiled kernel uses symbolic `M`, `K`, and batch dimensions, so one
> compiled binary can serve multiple sizes with the same dtype, layout, tile,
> and epilogue configuration. The existing QuACK object cache stores the
> compiled binary. Optional runtime autotuning selects between a small set of
> tile, cluster, and scheduler-group configurations and remembers the result.
>
> Supported operations are:
>
> ```text
> D = A @ A.mT
> D = alpha * A @ A.mT + beta * C    # C must also be symmetric
> ```
>
> ## Example
>
> ```python
> from torch._vendor.quack.gemm_symmetric import gemm_symmetric
>
> x = torch.randn(4096, 14336, device="cuda", dtype=torch.bfloat16)
> gram = torch.empty(4096, 4096, device="cuda", dtype=torch.bfloat16)
> gemm_symmetric(x, gram, autotune=True)
> torch.testing.assert_close(gram, x @ x.T)
>
> update = torch.empty_like(gram)
> gemm_symmetric(
>     gram,
>     update,
>     C=gram,
>     alpha=2.0315,
>     beta=-4.775,
>     autotune=True,
> )
> ```
>
> Unsupported architectures, dtypes, layouts, or shapes raise before launch;
> later integration commits decide when to select this kernel versus GEMM.
>
> ## Benchmark results
>
> Development measurements used BF16 on an NVIDIA GB200 and compare against
> the ordinary cuBLAS GEMM used for `A @ A.T`:
>
> | `A` shape | symmetric kernel | cuBLAS | speedup |
> |---|---:|---:|---:|
> | `(4096, 11008)` | 0.143 ms | 0.256 ms | 1.79x |
> | `(4096, 14336)` | 0.182 ms | 0.294 ms | 1.62x |
> | `(8192, 28672)` | 1.328 ms | 2.450 ms | 1.85x |
>
> The measured outputs were exactly symmetric, with maximum error versus the
> FP32 reference of approximately 0.004. These are development microbenchmarks,
> not performance guarantees.
>
> ## Test Plan
>
> ```text
> CUDA_VISIBLE_DEVICES=0 TORCHINDUCTOR_CACHE_DIR=/tmp/codex-pytorch2-final-symmetric-cache .venv/bin/python test/inductor/test_symmetric_mm.py -v
> CUDA_VISIBLE_DEVICES=0,1 .venv/bin/python -c 'import torch; from torch._vendor.quack.gemm_symmetric import gemm_symmetric; torch.cuda.set_device(0); x=torch.randn(512,1024,device="cuda:1",dtype=torch.bfloat16); out=torch.empty(512,512,device="cuda:1",dtype=torch.bfloat16); gemm_symmetric(x,out); torch.cuda.synchronize(1); torch.testing.assert_close(out,x@x.T); assert torch.cuda.current_device()==0'
> source .venv/bin/activate && lintrunner -a
> ```
>
> Authored with an AI assistant.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by mlazos** [comment]
## AI-Review

> *Automated multi-agent review (2 Claude + 2 Codex + GPT-5.6 verification). This is AI-generated — please verify findings manually.*

---

I’ll use the `pr-review` skill, as required by the repository instructions, and independently trace each proposed issue against the supplied change and relevant local implementation details.

**Comment by drisspg** [comment]
@claude fable is there anything that looks strange our out of the ordinary here

**Review by drisspg — APPROVED** [review]
stamp
